src/garage/replay_buffer/her_replay_buffer.py

```diff
@@ -2,145 +2,175 @@
 
 See: https://arxiv.org/abs/1707.01495.
 """
-import inspect
+import copy
 
 import numpy as np
 
-from garage.replay_buffer.replay_buffer import ReplayBuffer
+from garage.replay_buffer.path_buffer import PathBuffer
 
 
-def make_her_sample(replay_k, reward_fun):
-    """Generate a transition sampler for HER ReplayBuffer.
+class HerReplayBuffer(PathBuffer):
+    """Replay buffer for HER (Hindsight Experience Replay).
+
+    It constructs hindsight examples using future strategy.
 
     Args:
-        replay_k (float): Ratio between HER replays and regular replays
+        replay_k (int): Number of HER transitions to add for each regular
+            Transition. Setting this to 0 means that no HER replays will
+            be added.
         reward_fun (callable): Function to re-compute the reward with
-            substituted goals
+            substituted goals.
+        capacity_in_transitions (int): total size of transitions in the buffer.
+        env_spec (garage.envs.EnvSpec): Environment specification.
+    """
 
-    Returns:
-        callable: A function that returns sample transitions for HER.
+    def __init__(self, replay_k, reward_fun, capacity_in_transitions,
+                 env_spec):
+        self._replay_k = replay_k
+        self._reward_fun = reward_fun
+        self._env_spec = env_spec
 
-    """
-    future_p = 1 - (1. / (1 + replay_k))
+        if not float(replay_k).is_integer() or replay_k < 0:
+            raise ValueError('replay_k must be an integer.')
+        super().__init__(capacity_in_transitions)
+
+    def _sample_her_goals(self, path, transition_idx):
+        """Samples HER goals from the given path.
 
-    def _her_sample_transitions(episode_batch, sample_batch_size):
-        """Generate a dictionary of transitions.
+        Goals are randomly sampled starting from the index after
+        transition_idx in the given path.
 
         Args:
-            episode_batch (dict): Original transitions which
-                transitions[key] has shape :math:`(N, T, S^*)`.
-            sample_batch_size (int): Batch size per sample.
+            path (dict[string:ndarray]): A dict containing the transition
```

Comment:
isn't the syntax `Dict[string, np.ndarray]`?

Reply:
Fixed